app/utils/transformer/PageTransformer.scala

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils.transformer
+
+import models.UserAnswers
+import pages.QuestionPage
+import play.api.libs.json.{Reads, Writes}
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+trait PageTransformer {
+
+  type DomainModelType
+  type ExtractedTypeInDepartureData
+
+  def transformFromDeparture(
+    userAnswers: UserAnswers,
+    extractDataFromDepartureData: UserAnswers => Seq[ExtractedTypeInDepartureData],
+    generatePageAndValues: Seq[ExtractedTypeInDepartureData] => Seq[(QuestionPage[DomainModelType], DomainModelType)]
+  )(implicit writes: Writes[DomainModelType], reads: Reads[DomainModelType]): Future[UserAnswers] = {
+    val dataFromDepartureData = extractDataFromDepartureData(userAnswers)
+    val pageAndValueList      = generatePageAndValues(dataFromDepartureData)
+
+    pageAndValueList
+      .foldLeft(Try(userAnswers)) {
+        (accTry, pageAndValue) =>
+          accTry.flatMap(_.set(pageAndValue._1, pageAndValue._2))
+      }
+      .asFuture
+  }
+
+  def transformFromDepartureWithRefData(
+    userAnswers: UserAnswers,
+    fetchReferenceData: () => Future[Seq[DomainModelType]],
+    extractDataFromDepartureData: UserAnswers => Seq[ExtractedTypeInDepartureData],
+    generatePageAndValues: (Seq[ExtractedTypeInDepartureData], Seq[DomainModelType]) => Seq[(QuestionPage[DomainModelType], DomainModelType)]
```

Comment:
minor: Maybe personal preference, but could we have a custom dataType over the tuple?